<a href="https://colab.research.google.com/github/Gayatri1601R/IPSVIP-Task-2/blob/main/Code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**InfoPillar Solution PVT LTD**

**Name** : Gayatri Ganesh Rajopadhye

**Task-2** : Fake News Detection Project

**Importing** **Libraries**

In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

**Loading Dataset**

In [ ]:
dataset = pd.read_csv('/content/news.csv')
dataset

Unnamed: 0  ... label
0           8476  ...  FAKE
1          10294  ...  FAKE
2           3608  ...  REAL
3          10142  ...  FAKE
4            875  ...  REAL
...          ...  ...   ...
6330        4490  ...  REAL
6331        8062  ...  FAKE
6332        8622  ...  FAKE
6333        4021  ...  REAL
6334        4330  ...  REAL

[6335 rows x 4 columns]

**Summarise The Dataset**

In [ ]:
print("No. of Columns and Rows :- ",dataset.shape)

No. of Columns and Rows :-  (6335, 4)


In [ ]:
dataset.head()

Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]

**Labels from Dataset**

In [ ]:
labels=dataset.label
print(labels)

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object


**Splitting the Dataset for Training and Testing**

In [ ]:
x_train,x_test,y_train,y_test =train_test_split(dataset['text'], labels, test_size=0.2, random_state=7)

In [ ]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

**Fit and Transform the Vectorizer on the train set and transform the Vectorizer on the Test Set**

In [ ]:
tfidf_train=tfidf_vectorizer.fit_transform(x_train)
tfidf_test=tfidf_vectorizer.transform(x_test)

**Initialize a PassiveAggressiveClassifier**

In [ ]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac = pac.fit(tfidf_train,y_train)

**Predict on the Test set and Calculate Accuracy**

In [ ]:
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 93.05%


**The Accuracy is 93.05% .**

In [ ]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[593,  45],
       [ 43, 586]])

**In this model, we have 593 True Positives, 586 True Negatives, 43 False** **Positives and 45 False Negatives.**

**In SUMMARY, we learned to detect fake news with Python. We took a political** **dataset, implemented a TfidfVectorizer, initialized a** **PassiveAggressiveClassifier and fitted our model. We ended up obtaining an** **accuracy of 93.05% in magnitude.**

**Thank You**  